Wave variance

huwb opened the pull request:
**Status**: Ready for review. I have arranged these changes into logical commits so reviewing by-commit is a good option.

**Overview**:
Wave variance is statistical measure of how many waves there are in each slice of the displacement cascade. When sampling a low LOD of the waves, there are a lot of small wavelengths missing. The variance is a measure of the cumulative amount of missing content for a given slice. This measure can be used to compensate for the missing data. The killer use case of this is the foam at wave crests - right now we dont get foam at all in the mid-to-background as the pinchy waves are missing from the lower lods. The variance is used to compensate for this missing data and foam is now generated reliably at all LODs.

The variance is stored in the W channel of the displacement texture. Variance can be spatially varying so its put into the textures rather than factored out. This channel used to be used for SSS. **SSS has been moved away from this channel and put back into the ocean shader which computes it from surface pinch**. This is the way it used to be originally, and the results are very comparable.

Variance can in theory be useful for accounting for missing waves when approximating SSS, and also for driving surface material roughness in the distance. I didnt have a good fail case for the former, so i've left it disabled for now. I had a quick stab at the latter and it did work, but needs more love to make it mergeable, and it wasnt clear that driving it procedurally was a good idea rather than setting it manually, and this experiment is parked in `feature/variance-roughness`.

This only works with the new gerstner system. For the old gerstner system, the behaviour is unchanged from what it was before (the variance measure will just come through as 0 and have no impact).

**Testing**:
I have set up main.unity with the new gerstner component so can be used to test the changes.

daleeidd reviewed (approved):
Huge improvement. Only minor feedback but can be skipped.

There are some possible changes to SSS pinch but I will discuss on Discord as it is not really what this PR is about.

daleeidd commented on crest/Assets/Crest/Crest/Shaders/OceanHelpersNew.hlsl:
```diff
@@ -32,26 +35,48 @@ float2 IDtoUV(in float2 i_id, in float i_width, in float i_height)
+	{
+		// Don't use variance - yet
+		// float varianceBeforeThisCascade = data.w;
+		float4 du = float4(disp_x.xz, disp_z.xz) - disp.xzxz;
+		float det = (du.x * du.w - du.y * du.z) / (i_texelSize * i_texelSize);
```
We could use the built-in [determinate](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-determinant) function? Code would be:

```shader
float2x2 du = float2x2(disp_x.xz, disp_z.xz) - disp.xzxz;
float det = determinant(du) / (i_texelSize * i_texelSize);
```

Results looked the same.

I don't think there are any disadvantages. My understanding is that built-in function always perform better than custom ones (I read this during on of AMD's performance presentations if I recall correctly). I tried to avoid casting too since that can be another cost. I haven't measured this though.

Let me know what you think.

huwb replied:
good shout, thanks! i made it a bit more maths'y as well, and baked in the multipliers to clean it up

daleeidd commented on crest/Assets/Crest/Crest/Shaders/OceanHelpersNew.hlsl:
```diff
@@ -32,26 +35,48 @@ float2 IDtoUV(in float2 i_id, in float i_width, in float i_height)
 	}
+
+	// SSS - based off pinch
+	{
+		// Don't use variance - yet
+		// float varianceBeforeThisCascade = data.w;
+		float4 du = float4(disp_x.xz, disp_z.xz) - disp.xzxz;
+		float det = (du.x * du.w - du.y * du.z) / (i_texelSize * i_texelSize);
+		io_sss += i_wt * saturate( .03 * (20.0 - (det * 4.0 /*- varianceBeforeThisCascade * 20.0*/)) );
+	}
```
This could be wrapped in *_SUBSURFACESCATTERING_ON* to be explicit. But I think this should be stripped from the shader if that is not true anyway since the value would never be used.

huwb replied:
makes sense, done

huwb commented on crest/Assets/Crest/Crest/Shaders/Resources/UpdateFoam.compute:
```diff
@@ -83,9 +85,9 @@ void UpdateFoam(uint3 id : SV_DispatchThreadID)
 	float4 du = float4(disp_x.xz, disp_z.xz) - disp.xzxz;
 	float det = (du.x * du.w - du.y * du.z) / (cascadeData._texelWidth * cascadeData._texelWidth);
-	foam += 5.0 * _SimDeltaTime * _WaveFoamStrength * saturate(_WaveFoamCoverage - det);
+	foam += 5.0 * _SimDeltaTime * _WaveFoamStrength * saturate( _WaveFoamCoverage - det + foamBase*0.7 );
```
done

huwb commented on crest/Assets/Crest/Crest/Shaders/OceanHelpersNew.hlsl:
```diff
@@ -32,26 +35,48 @@ float2 IDtoUV(in float2 i_id, in float i_width, in float i_height)
+		// float varianceBeforeThisCascade = data.w;
+		float4 du = float4(disp_x.xz, disp_z.xz) - disp.xzxz;
+		float det = (du.x * du.w - du.y * du.z) / (i_texelSize * i_texelSize);
+		io_sss += i_wt * saturate( .03 * (20.0 - (det * 4.0 /*- varianceBeforeThisCascade * 20.0*/)) );
```
done

daleeidd replied:
Thanks

huwb commented:
@daleeidd  addressed feedback so far

the SSS is different from the old SSS. i tried to chill it out a bit but had limited success. i think your suggestion is probably needed - computing an averaged / lower lod. Perhaps it could be done in the vert shader.

if you agree i think we can merge this. ill be keen to see if the results hold up in URP/HDRP

daleeidd commented:
Sounds good. And agreed to merging.
